walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,103 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.ArrayList;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> timesForMeeting = new ArrayList<TimeRange>();
+
+    // Meeting is too long  
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return timesForMeeting;
+    }
+ 
+    // Assume optional attendees can't be included until
+    // slot is found
+    boolean optionalsIncluded = false;
+
+    // Begin search at start of day
+    int beginSlot = TimeRange.START_OF_DAY;
+
+    // Keep track of time slots for optional attendees
+    int beginSlotOptional = beginSlot;
+
+    for (Event event : events) {
+      // Check for conflicts
+      boolean noConflict = checkForConflict(event, request.getAttendees());
+      boolean noOptionalConflict = checkForConflict(event, request.getOptionalAttendees());
+
+      // If optionals can be included do original search method
+      if (optionalsIncluded) {
```

Comment:
In some of this your reaching a fairly deep level of nesting. You should look over this and consider how you can logically refactor flow to avoid this deep nesting.

Reply:
Might be worth drawing a truth table, btw. That's fairly useful for complex conditionals such as this.

Reply:
I figured out how to improve the logic and substantially reduce the nesting. I think I just needed a fresh go at this problem.